lib/internal/dns/compat.js

```diff
@@ -0,0 +1,249 @@
+'use strict';
+
+const {
+  ArrayPrototypeFilter,
+  ArrayPrototypeMap,
+  ArrayPrototypeSort,
+  JSONParse,
+  StringPrototypeReplace,
+  Symbol,
+} = primordials;
+
+const {
+  DNSWrap,
+  GetNameInfoReqWrap,
+  getnameinfo,
+} = internalBinding('dns');
+const { Buffer } = require('buffer');
+const { dnsException } = require('internal/errors');
+const { AsyncResource } = require('async_hooks');
+
+const kResource = Symbol('kResource');
+const kOnComplete = Symbol('kOnComplete');
+
+class ReqWrap {
+  constructor(name) {
+    this[kResource] = new AsyncResource(name);
```

Comment:
Could the `ReqWrap` itself just extend from `AsyncResource`?

Reply:
it could, though that would add the AsyncResource api to the req wrap, which is returned from all the various api calls to the user.